### ISSUE TITLE
[question] conanfile.py: license attribute and conandata

### ISSUE THREAD

**Issue by Johnnyxy** [conan-io/conan#7201]
Hi there,
I am fiddling around with 3rd party packages to provide them in my Conan environment.
Sometimes some 3rd party libraries do not have an appropriate license file (e.g. COPYING, LICENSE, ...) packed. So I have to add a license file as an extra to my recipe. For this I am using the [conan_data](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#conan-data) feature with the _conan_data.yml_.

As the license could change between versions of repacked 3rd party software I tried to use the conan_data feature with the [license attribute](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#license). 

Something like this:
```python
class package(ConanFile):
    ...
    license = conan_data["license"]
    ...
```

This does not work as the conan_data-object is a part of a conanfile-instance.

The [name attribute](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#name) has a function [set_name](https://docs.conan.io/en/latest/reference/conanfile/methods.html#set-name-set-version) which to change the name dynamically. This would be helpful with license (or others) too.

Best regards

**Comment by uilianries** [comment]
Hi @Johnnyxy !

Please, take a look on PR https://github.com/conan-io/conan-center-index/pull/1855/files

It has a similar case, you can use the same idea.

**Comment by uilianries** [comment]
For dynamic license name, you can use:

```
...
license = "BSD-2-Clause"

def configure(self):
    if tools.Version(self.version) > "1":
        self.license = "BSD-3-Clause"
```

When running `conan inspect` you will see BSD-2-Clause, because it's the static attribute value, but when running `conan info` to the specific package reference, you will see the correct version.

In my experience, it's very rare case when a project change its license, I saw it a couple of times.

**Comment by Johnnyxy** [comment]
Hi @uilianries,
yes normally this is very rare. But for example the PlantUML tool can be downloaded with different licenses ([see here](https://plantuml.com/download)). In the case of my recipe when someone adds a new version with a different license or the selected license gets dropped and some other has to be used I would have to provide a different license string. I was thinking about preparing the recipe for such a case.

But it is a minor problem. Your input is helpful and I will use that.
Thanks.

**Comment by uilianries** [comment]
@Johnnyxy indeed your case is very particular and adding a helper for license would help, but it's a new feature for a rare case ...

Cool that worked for you!

**Comment by memsharded** [comment]
Hi @Johnnyxy 

I am not sure if it is a good idea, or maybe a bit of abuse. But the new ``init()`` method, designed for python inheritance, should probably work to define the license. This is called just after the conan_data is initialized, and will be called even for ``conan inspect``.

**Comment by Johnnyxy** [comment]
Hm, ah yes, the _init_ method I forgot. Nowadays our IT department does not update Conan that often anymore so I came not in contact with some new features. It will have a look into it.

@memsharded 
This would mean that I could reinitialize every class member. But the static value still exists in the recipe:
```python
class TestClass(ConanFile):
    license = "MIT"

    def init(self):
        self.license = "GPL3"

print(TestClass.license) # "MIT"
print(TestClass().license) # "GPL3"
```

Is Conan now (maybe still in progress) redesigned to create an instance of a recipe on every access (like conan install, conan create, conan *whatever*, ... instead of accessing the values statically?
This would mean one could always decide what values will be set for license, name, channel, settings, etc. dynamically. I did not look that much into the Conan code to determine if the static access is absolutely necessary (e.g. for building the dependency tree, etc.).
